yarn-project/simulator/src/client/client_execution_context.ts

```diff
@@ -141,19 +141,18 @@ export class ClientExecutionContext extends ViewDataOracle {
    * Return the note encrypted logs emitted during this execution.
    */
   public getNoteEncryptedLogs() {
-    // Do not return logs that have been chopped in the cache
-    const filteredLogs = this.noteEncryptedLogs.filter(l => this.noteCache.getLogs().includes(l));
-    return filteredLogs;
+    return this.noteEncryptedLogs;
   }
 
   /**
    * Sometimes notes can be chopped after a nested execution is complete.
    * This means finished nested executions still hold transient logs. This method removes them.
    * TODO(Miranda): is there a cleaner solution?
    */
-  public chopNoteEncryptedLogsFromNested() {
+  public chopNoteEncryptedLogs() {
     // Do not return logs that have been chopped in the cache
     const allNoteLogs = this.noteCache.getLogs();
+    this.noteEncryptedLogs = this.noteEncryptedLogs.filter(l => allNoteLogs.includes(l));
```

Comment:
Not sure how this fixes it? Looks like we are still returning logs chopped in later functions from this execution.
I mean, we probably don't even need to call `chopNoteEncryptedLogs` at all 😂  
This should be handled by a top level component (like kernel prover), because only it has the whole picture and will know what are chopped in the tx.
For example, for the tx below, log_0 will be returned from here and included in the execution result.
```
fn A() {
  fn B() {
    emit note_hash_0;
    emit log_0;
  }
  fn C() {
    emit nullifier_for_note_hash_0;
  }
}
```

Reply:
This is called in `private_execution` and the loop below this bit of code removes all logs from nested executions e.g. from `fn B` above. I added the line above because it used to be covered in `.getNoteEncryptedLogs` but you made a good point that it's confusing for a getter to remove logs itself.
If it's not called, it does cause errors in your example since the nested execution for `B` will still have the log and will be gathered later in `collectSortedNoteEncryptedLogs`

Reply:
Before this change:
```
  /**
   * Return the note encrypted logs emitted during this execution.
   */
  public getNoteEncryptedLogs() {
    const filteredLogs = this.noteEncryptedLogs.filter(l => this.noteCache.getLogs().includes(l));
    return filteredLogs;
  }

  /**
   * Sometimes notes can be chopped after a nested execution is complete.
   * This means finished nested executions still hold transient logs. This method removes them.
   * TODO(Miranda): is there a cleaner solution?
   */
  public chopNoteEncryptedLogs() {
    // Do not return logs that have been chopped in the cache
    const allNoteLogs = this.noteCache.getLogs();
    const chop = (thing: any) =>
      thing.nestedExecutions.forEach((result: ExecutionResult) => {
        if (!result.noteEncryptedLogs[0]?.isEmpty()) {
          // The execution has note logs
          result.noteEncryptedLogs = result.noteEncryptedLogs.filter(l => allNoteLogs.includes(l));
        }
        chop(result);
      });
    chop(this);
  }
```
After this change:
```
  /**
   * Return the note encrypted logs emitted during this execution.
   */
  public getNoteEncryptedLogs() {
    return this.noteEncryptedLogs;
  }

  /**
   * Sometimes notes can be chopped after a nested execution is complete.
   * This means finished nested executions still hold transient logs. This method removes them.
   * TODO(Miranda): is there a cleaner solution?
   */
  public chopNoteEncryptedLogs() {
    // Do not return logs that have been chopped in the cache
    const allNoteLogs = this.noteCache.getLogs();
    this.noteEncryptedLogs = this.noteEncryptedLogs.filter(l => allNoteLogs.includes(l)); //Moved from above fn
    const chop = (thing: any) =>
      thing.nestedExecutions.forEach((result: ExecutionResult) => {
        if (!result.noteEncryptedLogs[0]?.isEmpty()) {
          // The execution has note logs
          result.noteEncryptedLogs = result.noteEncryptedLogs.filter(l => allNoteLogs.includes(l));
        }
        chop(result);
      });
    chop(this);
  }
``

Reply:
Oh I didn't read the enture chop function carefully. Now I see that it's changing the result for all the nested calls. So yeah you are right this should remove all the chopped logs correctly. Not sure I like this approach though 😅 But can refactor it later!

Reply:
I agree :/ I was thinking of refactoring the call to `collectSortedNoteEncryptedLogs`, but at that level we don't have a lot of info about chopped logs. Even though now we use `note_hash_counter` rather than `note_hash` I can't link ts logs to them and remove accordingly. I also can't compare to the public inputs, because we may just have a single `noteEncryptedLogsHash`. Will think on it!

Reply:
Maybe can do something similar as you've done in build_private_kernel_reset_outputs.ts?

Reply:
I would but I'm not sure where to put it - when we add the logs to the tx, there are often no `NoteLogHash`es because if we go straight to rollup, the kernel output gives `CombinedAccumulatedData`, which has the final logs hash only.